benches/benches/state.rs

```diff
@@ -0,0 +1,322 @@
+use criterion::{
+    criterion_group,
+    criterion_main,
+    Criterion,
+};
+use fuel_core::database::{
+    storage::{
+        ContractsStateMerkleData,
+        ContractsStateMerkleMetadata,
+        SparseMerkleMetadata,
+    },
+    vm_database::VmDatabase,
+};
+use fuel_core_storage::{
+    InterpreterStorage,
+    StorageAsMut,
+};
+use fuel_core_types::{
+    fuel_merkle::sparse::MerkleTree,
+    fuel_tx::Bytes32,
+    fuel_types::ContractId,
+};
+use rand::{
+    rngs::StdRng,
+    thread_rng,
+    Rng,
+    SeedableRng,
+};
+use std::iter;
+
+fn setup(db: &mut VmDatabase, contract: &ContractId, n: usize) {
+    let mut rng_keys = thread_rng();
+    let gen_keys = || -> Bytes32 { rng_keys.gen() };
+    let state_keys = iter::repeat_with(gen_keys).take(n);
+
+    let mut rng_values = thread_rng();
+    let gen_values = || -> Bytes32 { rng_values.gen() };
+    let state_values = iter::repeat_with(gen_values).take(n);
+
+    // State key-values
+    let state_key_values = state_keys.zip(state_values);
+
+    // Insert the key-values into the database while building the Merkle tree
+    let tree: MerkleTree<ContractsStateMerkleData, _> =
+        MerkleTree::from_set(db, state_key_values).expect("Failed to setup DB");
+
+    // Create the relevant Merkle metadata generated for the key values
+    let root = tree.root();
+    let metadata = SparseMerkleMetadata { root };
+
+    let db = tree.into_storage();
+    db.storage_as_mut::<ContractsStateMerkleMetadata>()
+        .insert(contract, &metadata)
+        .expect("Failed to create Merkle metadata");
+}
+
+fn state_single_contract(c: &mut Criterion) {
+    let mut rng = StdRng::seed_from_u64(0xF00DF00D);
+    let state: Bytes32 = rng.gen();
+    let value: Bytes32 = rng.gen();
+
+    let mut group = c.benchmark_group("state single contract");
+
+    group.bench_function("insert state with 0 preexisting entries", |b| {
+        let mut db = VmDatabase::default();
+        let contract: ContractId = rng.gen();
+        setup(&mut db, &contract, 0);
+        b.iter_custom(|iters| {
+            let start = std::time::Instant::now();
+            for _ in 0..iters {
+                db.merkle_contract_state_insert(&contract, &state, &value)
```

Comment:
We override the same storage key here each time, and maybe it is faster than inserting a new value. It would be nice to use the database without this entry.

I suggest creating a database transaction before each insertion, measuring the time for insertion, and dropping the database transaction. It should be more accurate.